buildkite/bazel_auto_sheriff.py

```diff
@@ -0,0 +1,643 @@
+            "annotate",
+        ] + (["--append"] if append else []) + [
+            f"--context={context}",
+            f"--style={style}",
+            f"\n{info_str}\n",
+        ],
+        print_output = False,
+    )
+
+
+# The class to fetch build result and do the analyzing
+# The build result are stored in main_result and downstream_result, their structures are like:
+# 1. An example of downstream _result
+# {
+#     "build_number": 1396,
+#     "bazel_commit": "83c5bb00b59137effd976e4f510f74613d85ee1c",
+#     "commit": "e2b6e2cffc79b89cf0840499c30e3818aaa9ee4b",
+#     "state": "passed",
+#     "tasks": {
+#         "ubuntu1604": {
+#             "id": "ba3b744c-536d-4f7b-ad9a-c2edd8839ba1",
+#             "name": "Android Studio Plugin (:ubuntu: 16.04 (OpenJDK 8))",
+#             "state": "passed",
+#             "web_url": "https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/1396#ba3b744c-536d-4f7b-ad9a-c2edd8839ba1"
+#         }
+#     }
+# }
+# 2. An example of main_result
+# {
+#     "build_number": 6299,
+#     "commit": "e2b6e2cffc79b89cf0840499c30e3818aaa9ee4b",
+#     "last_green_commit": "e2b6e2cffc79b89cf0840499c30e3818aaa9ee4b",
+#     "state": "passed",
+#     "tasks": {
+#         "ubuntu1604": {
+#             "id": "1333fa60-c26f-45dd-9169-eeb9a7cc9add",
+#             "name": ":ubuntu: 16.04 (OpenJDK 8)",
+#             "state": "passed",
+#             "web_url": "https://buildkite.com/bazel/android-studio-plugin/builds/6299#1333fa60-c26f-45dd-9169-eeb9a7cc9add"
+#         }
+#     }
+# }
+class BuildInfoAnalyzer(threading.Thread):
+
+    success_log_lock = threading.Lock()
+    success_log = []
+
+    def __init__(self, project, pipeline, downstream_result):
+        threading.Thread.__init__(self)
+        self.project = project
+        self.pipeline = pipeline
+        self.downstream_result = downstream_result
+        self.main_result = None
+        self.client = BuildkiteClient(BUILDKITE_ORG, pipeline)
+        self.analyze_log = [COLORS["HEADER"] + f"Analyzing {self.project}: " + COLORS["ENDC"]]
+        self.broken_by_infra = False
+
+
+    def __get_main_build_result(self):
+        build_info_list = self.client.get_build_info_list([
+            ("branch", "master"),
+            ("page", "1"),
+            ("per_page", "1"),
+            ("state[]", "failed"),
+            ("state[]", "passed"),
+        ])
+        if len(build_info_list) == 0:
+            error = f"Cannot find finished build for pipeline {self.pipeline}, please try to rerun the pipeline first."
+            self.__log("SERIOUS", error)
+            raise BuildkiteException(error)
+        main_build_info = build_info_list[0]
+
+        self.main_result = {}
+        self.main_result["commit"] = main_build_info["commit"]
+        self.main_result["build_number"] = main_build_info["number"]
+        job_infos = filter(lambda x: x != None, [extract_job_info_by_key(job) for job in main_build_info["jobs"]])
```

Comment:
more efficient if you use a generator expr instead of a list comprehension (just replace brackets with parentheses).

Reply:
Thanks!